### PR TITLE
EHN: data catalog unit attributes

### PR DESCRIPTION
**Pull request by Tjalling-dejong**
See issue #327

**Done**
- I implemented the functionality that allows for setting the attributes of variables in RasterDatasetAdapter.  The datacatalog allows an attrs argument that should refer to the variables in the dataset and can be given attribute data, for instance unit or long name of the variable.  
- Implement the same functionality for the GeoDatasetAdapter
- Implement the same functionality for the DataframeAdapter
- And GeoDataframeAdapter

### CODE REVIEW

File: hydromt/data_adapter/data_adapter.py [hunk 1]
```diff
@@ -140,16 +181,19 @@ def __init__(
...
         self.nodata = nodata
         self.rename = rename
         self.unit_mult = unit_mult
         self.unit_add = unit_add
-        self.zoom_levels = zoom_levels
         # meta data
         self.meta = {k: v for k, v in meta.items() if v is not None}
+        # variable attributes
+        # for var in self.var:
+        #     self[var].attrs.update(**attrs[var])
```
Comment (savente93) [inline hydromt/data_adapter/data_adapter.py:195]: Can we remove this commented code?

File: tests/test_data_adapter.py [hunk 2]
```diff
@@ -161,6 +215,41 @@ def test_geodataset(geoda, geodf, ts, tmpdir):
...
+
+
+def test_geodataset_unit_conversion(artifact_data: DataCatalog):
+    gtsm_geodataarray = artifact_data.get_geodataset("gtsmv3_eu_era5")
+    gtsm_dict = {"gtsmv3_eu_era5": artifact_data.sources["gtsmv3_eu_era5"].to_dict()}
+    gtsm_dict["gtsmv3_eu_era5"].update(dict(unit_mult=dict(waterlevel=1000)))
+    datacatalog = DataCatalog()
+    datacatalog.from_dict(gtsm_dict)
+    gtsm_geodataarray1000 = datacatalog.get_geodataset("gtsmv3_eu_era5")
+    assert gtsm_geodataarray1000.equals(gtsm_geodataarray * 1000)
```
Comment (savente93) [inline tests/test_data_adapter.py:240]: If we multiply like this, should the unit name name also be adjusted? (no is a perfectly acceptable answer here, I'm genuinely not sure)

File: hydromt/vector.py [hunk 3]
```diff
@@ -29,6 +29,7 @@ def __init__(self, xarray_obj):
         """Initialize a new object based on the provided xarray_obj."""
         super(GeoBase, self).__init__(xarray_obj)
         self._geometry = None
+        self.nodata = np.nan  # placeholder nodata variable TODO
```
Comment (savente93) [inline hydromt/vector.py:32]: Is it okay to keep the placeholder like this, or should we decide on a value?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -12,11 +12,11 @@ repos:
...
-  - repo: https://github.com/jumanjihouse/pre-commit-hook-yamlfmt
-    rev: 0.2.1
-    hooks:
-      - id: yamlfmt
-        args: [--mapping, '2', --sequence, '4', --offset, '2']
+  # - repo: https://github.com/jumanjihouse/pre-commit-hook-yamlfmt
+  #   rev: 0.2.1
+  #   hooks:
+  #     - id: yamlfmt
+  #       args: [--mapping, '2', --sequence, '4', --offset, '2']
```
Comment (savente93) [inline .pre-commit-config.yaml:19]: Be sure to fix this merge conflict before you merge. pulling main should be ok.

**Review by savente93 — COMMENTED** [review]
Very nice work Tjalling, code looks to be very well tested. Just a few comments about polish, but most of it looks good. Will be happy to approve once comments are resolved.

File: hydromt/data_adapter/data_adapter.py [hunk 5]
```diff
@@ -174,6 +218,10 @@ def to_dict(self):
         for k, v in vars(self).items():
             if k in ["name", "catalog_name"]:
                 continue  # do not add these identifiers
+            if (
+                k == "driver_kwargs"
+            ):  # Prevent driver_kwargs name from ending up in the data catalog
+                k = "kwargs"
```
Comment (DirkEilander) [inline hydromt/data_adapter/data_adapter.py:224]: Why is this needed? basically have decided that "driver_kwargs" replaces the old "kwargs" right?

**Review by DirkEilander — CHANGES_REQUESTED** [review]
Indeed nice work @Tjalling-dejong! The code looks good to me. Regarding the docs I have two remarks
- I noticed some errors we made copy-pasting the docstrings for the get_data methods. Some of the docstrings info applies to other DataAdapters. These should be fixed. 
- Could you still add a description of the new `attrs` keyword
- Could you add a note to `driver_kwargs` and `attrs` that these keywords have been changed / added that from version 0.7.2

both changes re in docs/user_guide/data_prepare_cat.rst I think

File: tests/test_data_adapter.py [hunk 6]
```diff
@@ -161,6 +215,41 @@ def test_geodataset(geoda, geodf, ts, tmpdir):
...
+
+
+def test_set_nodata_geodataset(artifact_data: DataCatalog):
+    gtsm_dict = {"gtsmv3_eu_era5": artifact_data.sources["gtsmv3_eu_era5"].to_dict()}
+    gtsm_dict["gtsmv3_eu_era5"].update(dict(nodata=-99))
+    datacatalog = DataCatalog()
+    datacatalog.from_dict(gtsm_dict)
+    datacatalog.get_geodataset("gtsmv3_eu_era5")
+    # assert gtsm_geodataarray.vector.nodata == -99 TODO assertion cannot be made
+    # since set_noda for GeoDataArray.vector does not exist
```
[the remark this answers is not in the document]
Comment (Tjalling-dejong) [reply tests/test_data_adapter.py:250]: Yes, I made a new issue (#431) for it. I'll remove the comments.

File: hydromt/data_adapter/data_adapter.py (hunk 1 above)
Comment (Tjalling-dejong) [reply hydromt/data_adapter/data_adapter.py:195]: Yes, I removed it.

File: tests/test_data_adapter.py (hunk 2 above)
Comment (Tjalling-dejong) [reply tests/test_data_adapter.py:240]: In theory yes, but we are only testing the multiplication of the units. Changing the unit name is already covered in other tests.

File: tests/test_data_adapter.py [hunk 7]
```diff
@@ -232,26 +355,53 @@ def test_dataframe_time(df_time, tmpdir):
...
     )
     assert np.all(dfts5.columns == vars_slice)
 
 
+def test_dataframe_parse_nodata(artifact_data: DataCatalog, tmpdir):
+    hydro_reservoirs = artifact_data.get_geodataframe(
+        "hydro_reservoirs"
+    )  # This dataset has nodata set to -99
+    dataframe = pd.DataFrame(hydro_reservoirs.drop(columns="geometry"))
+    dataframe["Dam_height"] = -99
```
[the remark this answers is not in the document]
Comment (Tjalling-dejong) [reply tests/test_data_adapter.py:390]: It can have any name tbh, its only dummy data :)

File: hydromt/vector.py (hunk 3 above)
Comment (Tjalling-dejong) [reply hydromt/vector.py:32]: I removed it

File: .pre-commit-config.yaml (hunk 4 above)
Comment (Tjalling-dejong) [reply .pre-commit-config.yaml]: fixed it

**Review by DirkEilander — APPROVED** [review]
I've made a few final changes, mainly to the docs and tests as well as adding a nodata property to the GeoDataset class. (vector accessor). Can you have a look at my final changes and merge if you're happy @Tjalling-dejong?

File: hydromt/data_adapter/dataframe.py [hunk 8]
```diff
@@ -24,19 +25,18 @@ class DataFrameAdapter(DataAdapter):
 
     def __init__(
         self,
-        path,
-        driver=None,
-        filesystem="local",
-        crs=None,
```
Comment (DirkEilander) [inline hydromt/data_adapter/dataframe.py]: DataFrames do not have a CRS property as it does not have spatial data

**Pull request merged**